localstack/aws/serving/edge.py

```diff
@@ -15,6 +16,11 @@ def serve_gateway(bind_address, port, use_ssl, asynchronous=False):
     server = GatewayServer(gateway, port, bind_address, use_ssl)
     server.start()
 
+    # with the current way the infrastructure is started, this is the easiest way to shut down the server correctly
+    # FIXME: but the infrastructure shutdown should be much cleaner, core components like the gateway should be handled
+    #  explicitly by the thing starting the components, not implicitly by the components.
+    SHUTDOWN_HANDLERS.register(server.shutdown)
```

Comment:
We are using the `GatewayServer` now not only for the ASF gateway, but also for other purposes (like the `ProxyServer`). And there are lots of other server instances which should maybe be shut down properly.

Wouldn't it also be a viable solution to define that the lifecycle of every `Server` instance ends when LocalStack is shut down? Shouldn't basically all started servers be added to the shutdown handlers?

If not, we would have to be way more strict with the service provider's `on_before_stop` implementations (where all services would have to properly shut down all their started services).

Reply:
I feel this might be a good backup, but every service provider should do its own cleanup of resources independently. Especially when taking into account that, at some point, we might want to stop / restart services without stopping LocalStack itself. This would also simplify actions like `reset`.